python/srf/_pysrf/include/pysrf/edge_adapter.hpp

```diff
@@ -0,0 +1,247 @@
+            }
+
+            // Check if the sink is a py::object
+            if (reader_type == typeid(PyHolder) && writer_typei)
+            {
+                // TODO(MDD): To avoid a compound edge here, register an edge converter between OutputT and py::object
+                node::EdgeConnector<OutputT, PyHolder>::register_converter();
+
+                // Build the edge with the holder type
+                return srf::node::EdgeBuilder::ingress_for_source_type(source.source_type(), sink, ingress_handle);
+            }
+
+            // Check if both have been registered with pybind 11
+            if (writer_typei && reader_typei)
+            {
+                // TODO(MDD): Check python types to see if they are compatible
+
+                // Py types can be converted but need a compound edge. Build that here
+                // auto py_to_sink_edge = std::dynamic_pointer_cast<channel::Ingress<pybind11::object>>(
+                //    sink.ingress_for_source_type(typeid(pybind11::object)));
+                auto py_to_sink_edge = std::dynamic_pointer_cast<channel::Ingress<PyHolder>>(
+                    srf::node::EdgeBuilder::ingress_for_source_type(typeid(PyHolder), sink, ingress_handle));
+
+                auto source_to_py_edge = std::make_shared<node::Edge<OutputT, PyHolder>>(py_to_sink_edge);
+
+                LOG(WARNING) << "WARNING: A slow edge connection between C++ nodes '" << source.source_type_name()
+                             << "' and '" << sink.sink_type_name()
+                             << "' has been detected. Performance between "
+                                "these nodes can be improved by registering an EdgeConverter at compile time. Without "
+                                "this, conversion "
+                                "to an intermediate python type will be necessary (i.e. C++ -> Python -> C++).";
+
+                return std::dynamic_pointer_cast<channel::IngressHandle>(source_to_py_edge);
+            }
+
+            // Otherwise return base which most likely will fail
+            return srf::node::EdgeBuilder::default_ingress_adapter_for_sink(source, sink, ingress_handle);
+        };
+    }
+};
+
+/**
+ * @brief Sources which inherit this object will automatically attempt to a pySRF adapter for their data type with the
+ * EdgeAdaptorRegistry
+ * @tparam SourceT Data type the inheriting source emits
+ */
+template <typename SourceT>
+struct AutoRegSourceAdapter
+{
+    static bool s_initialized;
+
+    AutoRegSourceAdapter()
+    {
+        // force register_adapter to be called by anyone who inherits.
+        auto _initialized = s_initialized;
+    }
+
+    static bool register_adapter()
+    {
+        if (!srf::node::EdgeAdapterRegistry::has_source_adapter(typeid(SourceT)))
+        {
+            std::type_index source_type = typeid(SourceT);
+            VLOG(2) << "Registering PySRF source adapter for: " << type_name<SourceT>() << " "
```

Comment:
We need an `enum` for the allowed values for vlog and consistent guidelines on what level to use.

Should probably open an issue for arch review

Reply:
Done

Reply:
done.